File: Structural/Facade/src/examplejava/VideoConversionFacade.java
package examplejava;

import examplejava.codec.Codec;
import examplejava.codec.CodecFactory;
import examplejava.codec.MPEG4CompressionCodec;
import examplejava.codec.OggCompressionCodec;

import java.io.File;

public class VideoConversionFacade {
    public File convertVideo(String fileName, String format) {
        System.out.println("VideoConversionFacade: conversion started.");
        VideoFile file = new VideoFile(fileName);
        Codec sourceCodec = CodecFactory.extract(file);
        Codec destinationCodec;
        if (format.equals("mp4")) {
            destinationCodec = new OggCompressionCodec();
        } else {
            destinationCodec = new MPEG4CompressionCodec();
        }
        VideoFile buffer = BitrateReader.read(file);
        VideoFile intermediateResult = BitrateReader.convert(buffer);
        File result = (new AudioMixer()).fix();
        System.out.println("VideoConversionFacade: conversion completed.");
        return result;
    }
}

File: Structural/Facade/src/examplekotlin/VideoConversionFacade.kt
package examplekotlin

import examplekotlin.codec.CodecFactory
import examplekotlin.codec.MPEG4CompressionCodec
import examplekotlin.codec.OggCompressionCodec
import java.io.File

object VideoConversionFacade {
    fun convertVideo(fileName: String, format: String): File {
        println("VideoConversionFacade: conversion started.")
        val videoFile = VideoFile(fileName)
        val sourceCodec = CodecFactory.extract(videoFile)
        val destinationCodec = when (format) {
            "mp4" -> OggCompressionCodec()
            else -> MPEG4CompressionCodec()
        }
        val buffer = BitrateReader.read(videoFile, sourceCodec)
        val intermediateResult = BitrateReader.convert(buffer, destinationCodec)
        val result: File = AudioMixer.fix(intermediateResult)
        println("VideoConversionFacade: conversion completed.")
        return result
    }
}